### PR TITLE
DM-13054: Add colorterm config support to jointcal

### PR DESCRIPTION
**Pull request by parejkoj**

### CODE REVIEW

File: python/lsst/jointcal/jointcal.py [hunk 1]
```diff
@@ -614,6 +609,67 @@ def _do_load_refcat_and_fit(self, associations, defaultFilter, center, radius,
...
+        if not skyCircle.refCat.isContiguous():
+            refCat = skyCircle.refCat.copy(deep=True)
+        else:
+            refCat = skyCircle.refCat
+
+        if applyColorterms:
+            try:
+                refCatName = refObjLoader.ref_dataset_name
+            except AttributeError:
+                # NOTE: we need this try:except: block in place until we've completely removed a.net support.
```
Comment (erykoff) [inline python/lsst/jointcal/jointcal.py]: Oh so close!

File: python/lsst/jointcal/jointcal.py [hunk 2]
```diff
@@ -614,6 +609,67 @@ def _do_load_refcat_and_fit(self, associations, defaultFilter, center, radius,
...
+                          filterName, refCatName)
+            colorterm = self.config.colorterms.getColorterm(
+                filterName=filterName, photoCatName=refCatName, doRaise=True)
+
+            refMag, refMagErr = colorterm.getCorrectedMagnitudes(refCat, filterName)
+            refCat[skyCircle.fluxField] = u.Magnitude(refMag, u.ABmag).to_value(u.nJy)
+            # TODO: I didn't want to use this, but I'll deal with it in DM-16903
+            refCat[skyCircle.fluxField+'Err'] = fluxErrFromABMagErr(refMagErr, refMag) * 1e9
+        else:
+            # TODO: need to scale these until RFC-549 is completed and refcats return nanojansky
```
Comment (erykoff) [inline python/lsst/jointcal/jointcal.py]: Are these two TODOs related?

File: tests/test_jointcal_hsc.py [hunk 3]
```diff
@@ -150,6 +152,31 @@ def test_jointcalTask_2_visits_simpleMagnitude(self):
...
+        self.config.photometryModel = "simpleMagnitude"
+        self.config.applyColorTerms = True
+        # The values in this library were taken from obs_subaru as of stack version 16.0
+        self.config.colorterms = ColortermLibrary(data={
+            "sdss*": ColortermDict(data={
+                'r': Colorterm(primary="r", secondary="i", c0=0.00231810, c1=0.01284177, c2=-0.03068248)
+            })
+        })
+        # Final chi2 should be different, but I don't have an a-priori reason
+        # to expect it to be larger or smaller.
```
Comment (erykoff) [inline tests/test_jointcal_hsc.py]: If the color terms are doing what they should be doing, and if the reference stars have a wide enough color range (which may not be the case!) then I would expect the chi2 to be slightly smaller.  But if the color corrections are subtle and it's not good statistics, then I can see how it could increase.

File: src/Associations.cc [hunk 4]
```diff
@@ -229,9 +213,7 @@ void Associations::collectRefStars(afw::table::SimpleCatalog &refCat, afw::geom:
         star->vxy = 0.;
 
         // Reject sources with non-finite fluxes and flux errors, and fluxErr=0 (which gives chi2=inf).
-        if (rejectBadFluxes &&
-            (!std::isfinite(defaultFlux) || !std::isfinite(defaultFluxErr) || defaultFluxErr == 0))
-            continue;
+        if (rejectBadFluxes && (!std::isfinite(flux) || !std::isfinite(fluxErr) || fluxErr == 0)) continue;
```
Comment (erykoff) [inline src/Associations.cc]: I don't know of all the ways in which a catalog could have "bad" fluxes, but I would replace the last term with `fluxErr <= 0` to avoid anything with a (clearly wrong) negative flux error.

File: python/lsst/jointcal/jointcal.py [hunk 5]
```diff
@@ -614,6 +609,67 @@ def _do_load_refcat_and_fit(self, associations, defaultFilter, center, radius,
...
+        fluxField : `str`
+            The name of the reference catalog flux field appropriate for ``filterName``.
+        """
+        skyCircle = refObjLoader.loadSkyCircle(center,
+                                               afwGeom.Angle(radius, afwGeom.radians),
+                                               filterName)
+
+        # Need memory contiguity to get reference filters as a vector.
+        if not skyCircle.refCat.isContiguous():
+            refCat = skyCircle.refCat.copy(deep=True)
```
Comment (erykoff) [inline python/lsst/jointcal/jointcal.py]: Okay, here's a continuity check, but does there need to be one in `pipe_tasks` as well?

File: python/lsst/jointcal/jointcal.py [hunk 6]
```diff
@@ -614,6 +609,67 @@ def _do_load_refcat_and_fit(self, associations, defaultFilter, center, radius,
 
         return result
 
+    def _load_reference_catalog(self, refObjLoader, center, radius, filterName,
+                                applyColorterms=False):
```
Comment (erykoff) [inline python/lsst/jointcal/jointcal.py]: Could this method be moved out of `jointcal` and into somewhere else (`pipe_tasks`, I guess?) where it could be a stand-alone method?  I think it could be generally useful (I'd be writing something almost exactly like it for `fgcmcal`), and the only reference to `self` seems to be for the logging.

File: python/lsst/jointcal/jointcal.py (hunk 2 above)
Comment (parejkoj) [reply python/lsst/jointcal/jointcal.py]: Yes, although they may not happen at the exact same time.

File: tests/test_jointcal_hsc.py (hunk 3 above)
Comment (parejkoj) [reply tests/test_jointcal_hsc.py]: In this case, I think we might be in the "not good statistics" situation: there are only two hsc visits going into this, and only a handful of ccds, and it's a "simple" model, so this is mostly just to demonstrate that it "does something".

File: python/lsst/jointcal/jointcal.py (hunk 5 above)
Comment (parejkoj) [reply python/lsst/jointcal/jointcal.py]: Probably. That would be ticket DM-14061.
